### PR TITLE
Add wandb.require() feature flag

### PR DESCRIPTION
**Pull request by raubitsj**
Description
-----------

Add ability to specify specify that a wandb script is going to use experimental features.

For example:
```py
import wandb
wandb.require("wandb-service")
```

This could indicate that a new experimental feature is being used in the script,  for example `wandb-service` feature could add a new top level function:
```py
import wandb
wandb.require("wandb-service")
service = wandb.service()
service.start()
```

Without the `require()` call, the function `service()` is hidden from the standard interface, likely `_service()`.   The `require()` call makes it available in the api.

This line becomes a no-op when the feature is released.

Versions can be specified which can optionally alert the user when an interface has changed in a new version of the library.
```py
wandb.require("wandb-service@beta")
```
(versions are thought of a lexicographically sorted, so beta > alpha2 > alpha)

Alternatives considered
-----------------------

```py
from wandb.future import __wandb_service__
```
This is known in python world, but it is more awkward to do versioning

Testing
-------

Unittests (100% coverage)

TODO:
- [x] Add exception for unsupported features

Future work: (in a PR that adds a feature)
- telemetry
- parse versions and create exception for mismatched requirements (asking for "beta" when "beta2" is out)
- add map for `supported` features, feature/version tuples
- add URL and doc page for feature information

### CODE REVIEW

File: tests/test_feature_user.py [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+
+
+def test_feature_good(user_test, feature_mock):
+    def mock_feature_test(self):
+        wandb.feature_test = lambda x: x + 2
+
+    feature_mock("test", mock_feature_test)
+    wandb.use_feature("test")
+
+    assert wandb.feature_test(2) == 4
```
Comment (dannygoldstein) [inline tests/test_feature_user.py:70]: Can we add a test that a specific version of feature mock is picked up?

Comment (dannygoldstein) [reply tests/test_feature_user.py:70]: E.g., beta

File: wandb/sdk/wandb_feature.py [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+
+class _Features(object):
+    """Internal feature class"""
+
+    _features: List[str]
+
+    def __init__(self, features: str) -> None:
+        self._features = features.split(",")
+
+    def apply(self) -> None:
```
Comment (dannygoldstein) [inline wandb/sdk/wandb_feature.py:28]: Might be good to add some logic to handle improperly formatted feature names.

File: wandb/sdk/wandb_feature.py [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+    Raises:
+        wandb.errors.UseFeatureError: if not supported or other error
+    """
+    for v in args:
+        wandb.termwarn("use_feature() ignoring unsupported parameter: {}".format(v))
+
+    for k in kwargs:
+        wandb.termwarn(
+            "use_feature() ignoring unsupported named parameter: {}".format(k)
+        )
```
Comment (dannygoldstein) [inline wandb/sdk/wandb_feature.py:62]: Why not just remove these?

File: wandb/sdk/wandb_feature.py [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+
+    f = _Features(features=features)
+    f.apply()
```
Comment (dannygoldstein) [inline wandb/sdk/wandb_feature.py:65]: Is there a need for an internal class here? Might be a bit easier to read if we just copy paste the class contents here, unless the class is going to also be used elsewhere

File: tests/test_feature_user.py (hunk 1 above)
Comment (dannygoldstein) [reply tests/test_feature_user.py:70]: When there are multiple versions present

File: wandb/sdk/wandb_feature.py (hunk 3 above)
Comment (dannygoldstein) [reply wandb/sdk/wandb_feature.py:62]: And the associated `*args`, `**kwargs`

File: tests/test_feature_user.py (hunk 1 above)
Comment (raubitsj) [reply tests/test_feature_user.py:70]: no version support in this PR, it comes when we add a version. (listed as future work)

File: wandb/sdk/wandb_feature.py (hunk 2 above)
Comment (raubitsj) [reply wandb/sdk/wandb_feature.py:28]: We can, but it is handled, it wont match a lookup, worst that can happen is amessage saying feature KF*#@HLFA isnt supported.

Comment (dannygoldstein) [reply wandb/sdk/wandb_feature.py:28]: ok, no need in that case

File: wandb/sdk/wandb_feature.py (hunk 3 above)
Comment (raubitsj) [reply wandb/sdk/wandb_feature.py:62]: The idea is that feature could be extended, i want to move to this, so we can add more args but users of old libraries will get an indication that they are using a feature that doesnt exist in their older client version

File: wandb/sdk/wandb_feature.py (hunk 4 above)
Comment (raubitsj) [reply wandb/sdk/wandb_feature.py:65]: easier to test, easier to add functionally, encapsulate state (features which will grow over time)

Comment (dannygoldstein) [reply wandb/sdk/wandb_feature.py:65]: feel free to disregard this

**Review by dannygoldstein — APPROVED** [review]

**Pull request merged**